### PR TITLE
Updated C++/WinRT to 2.0.200316.3 and Min Windows to RS3+ (10.0.16299.0)

### PR DESCRIPTION
**Pull request by vmoroz**
We need to make the .Net Native work (Issue #3838). This change does not fix the issues we see, but gets us to the latest versions of frameworks we use and will enable us to solve in future. This PR includes the following changes:
- C++/WinRT is updated from 2.0.190730.2 to the latest 2.0.200316.3
  - It has all the latest C++/WinRT fixes and it should help us stay up-to-date with the C++/WinRT changes.
  - Changed Desktop ABI code and the related unit tests.
  - Removed use of [noexcept2] attribute - it does not work well with the new C++/WinRT version. Now it generates slightly inefficient code, but it must not affect any core functionality. In future we will be able to use the standard [noexcept] attribute after we start using the new version of Windows SDK.
- Minimum supported version of Windows is changed from RS2 (10.0.15063.0) to RS3 (10.0.16299.0)
  - It should enable us using the latest UWP .Net support that includes .Net Core version 2.1+ instead of .Net Core 1.0. E.g. we will be able to use Roslyn code generation tools.
  - SampleCppApp C++ project is changed to support newer version of .Net Native runtime.
- Fixed C# compilation warnings.

In future PRs we must:
- Change C++ App template to use the latest .Net Native runtime if it references .Net components.
- Use Roslyn code gen to replace reflection in .Net projects or fix the failing reflection techniques.

This PR may help addressing the #4031 issue.

###### Microsoft Reviewers: [Open in CodeFlow](http://wpcp.azurewebsites.net/CodeFlowProtocolProxy2.php?pullrequest=https://github.com/microsoft/react-native-windows/pull/4449)

### CODE REVIEW

File: vnext/Desktop/module.g.cpp [hunk 1]
```diff
@@ -0,0 +1,74 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+#include "pch.h"
```
Comment (NikoAri) [inline vnext/Desktop/module.g.cpp]: >pch [](start = 10, length = 3)

What generates this .g file?

File: vnext/Desktop/module.g.cpp [hunk 2]
```diff
@@ -0,0 +1,74 @@
...
+void* winrt_make_facebook_react_NativeTraceEventSource();
+
+bool __stdcall winrt_can_unload_now() noexcept
+{
+    if (winrt::get_module_lock())
+    {
+        return false;
+    }
+
+    winrt::clear_factory_cache();
```
Comment (NikoAri) [inline vnext/Desktop/module.g.cpp]: >clear_factory_cache(); [](start = 11, length = 22)

why is this needed here in getter function?

File: vnext/Desktop/module.g.cpp [hunk 3]
```diff
@@ -0,0 +1,74 @@
...
+    return true;
+}
+
+void* __stdcall winrt_get_activation_factory([[maybe_unused]] std::wstring_view const& name)
+{
+    auto requal = [](std::wstring_view const& left, std::wstring_view const& right) noexcept
```
Comment (NikoAri) [inline vnext/Desktop/module.g.cpp]: >requal [](start = 9, length = 6)

(nit) why not just spell out "AreEqual"?

File: vnext/Desktop/module.g.cpp [hunk 4]
```diff
@@ -0,0 +1,74 @@
...
+        return 0;
+    }
+
+#ifdef _WRL_MODULE_H_
+    return ::Microsoft::WRL::Module<::Microsoft::WRL::InProc>::GetModule().GetActivationFactory(static_cast<HSTRING>(classId), reinterpret_cast<::IActivationFactory**>(factory));
+#else
+    return winrt::hresult_class_not_available(name).to_abi();
+#endif
+}
+catch (...) { return winrt::to_hresult(); }
```
Comment (NikoAri) [inline vnext/Desktop/module.g.cpp]: Add comment that this function returns HRESULT?

Also, should there be some sort of Assert here?

File: vnext/Desktop/module.g.cpp (hunk 1 above)
Comment (vmoroz) [reply vnext/Desktop/module.g.cpp]: It is a "copy&paste" of file generated by C++/WinRT. For some reason, in this case it generates a file that has non-C++ identifiers and I had manually update it and make a copy.

File: vnext/Desktop/module.g.cpp (hunk 2 above)
Comment (vmoroz) [reply vnext/Desktop/module.g.cpp]: I do not know. It is generated by C++/WinRT.

File: vnext/Desktop/module.g.cpp (hunk 3 above)
Comment (vmoroz) [reply vnext/Desktop/module.g.cpp]: C++/WinRT code.

File: vnext/Desktop/module.g.cpp (hunk 4 above)
Comment (vmoroz) [reply vnext/Desktop/module.g.cpp]: I would prefer not to change this code. because it is generated by C++/WinRT.

**Comment by JunielKatarn** [comment]
Should this be ported back to `0.60-stable`?

**Comment by NickGerleman** [comment]
I don't think most of this exists in a usable form in 0.60, but even for 0.61 it would be breaking. Letting this show up in 0.62 seems reasonable to me.
